raiden/smart_contracts/TokenNetwork.sol

```diff
@@ -0,0 +1,697 @@
+    /// to the channel participant who calls the function.
+    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
+    /// @param additional_hash Computed from the message. Used for message authentication.
+    /// @param signature Partner's signature of the balance proof data.
+    function closeChannel(
+        uint256 channel_identifier,
+        uint64 nonce,
+        uint256 transferred_amount,
+        bytes32 locksroot,
+        bytes32 additional_hash,
+        bytes signature)
+        public
+    {
+        address partner_address;
+
+        // Close can be called only once
+        require(closing_requests[channel_identifier].settle_block_number == 0);
+
+        Channel storage channel = channels[channel_identifier];
+
+        // Only a participant may close the channel.
+        require(channel.participants[msg.sender].initialized);
+
+        // Store the closing request data
+        closing_requests[channel_identifier].closing_participant = msg.sender;
+        closing_requests[channel_identifier].settle_block_number = channel.settle_timeout + block.number;
+
+        // An empty value means that the closer never received a transfer, or
+        // he is intentionally not providing the latest transfer, in which case
+        // the closing party is going to lose the tokens that were transferred
+        // to him.
+        partner_address = recoverAddressFromSignature(
+            channel_identifier,
+            nonce,
+            transferred_amount,
+            locksroot,
+            additional_hash,
+            signature
+        );
+
+        // Signature must be from the channel partner
+        require(msg.sender != partner_address);
+
+        updateParticipantStruct(
+            channel_identifier,
+            partner_address,
+            nonce,
+            locksroot,
+            transferred_amount
+        );
+
+        ChannelClosed(channel_identifier, msg.sender);
+    }
+
+    /// @notice Called on a closed channel, the function allows the non-closing participant
+    /// to provide the last balance proof, which modifies the closing participant's state.
+    /// Can be called multiple times, only by the non-closing participant.
+    /// @param channel_identifier The channel identifier - mapping key used for `channels`.
+    /// @param nonce Strictly monotonic value used to order transfers.
+    /// @param transferred_amount Total amount of tokens transferred by the channel partner
+    /// to the channel participant who calls the function.
+    /// @param locksroot Root of the partner's merkle tree of all pending lock lockhashes.
+    /// @param additional_hash Computed from the message. Used for message authentication.
+    /// @param closing_signature Signature of the closing participant on the balance proof data.
+    function updateTransfer(
+        uint256 channel_identifier,
```

Comment:
the token_address is missing here too

Reply:
Same. 
I will add the `chain_id`.
I used the TokenNetwork contract address in `recoverAddressFromSignature`.

Reply:
you're right, I didn't look inside the `recoverAddressFromSignature` function.